controllers/datamovement_controller_test.go

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"context"
+	"os"
+
+	nnfv1alpha1 "github.com/NearNodeFlash/nnf-sos/api/v1alpha1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Data Movement Test" /*Ordered, (Ginkgo v2)*/, func() {
+	var dm *nnfv1alpha1.NnfDataMovement = nil
+	var srcPath, destPath string
+	var err error
+
+	// test is using `sleep 1` for datamovement, so add some padding
+	dmTestTimeout := 3
+
+	BeforeEach(func() {
+		srcPath, err = os.MkdirTemp("/tmp", "dm-test")
+		Expect(err).ToNot(HaveOccurred())
+
+		destPath, err = os.MkdirTemp("/tmp", "dm-test")
+		Expect(err).ToNot(HaveOccurred())
+
+		dm = &nnfv1alpha1.NnfDataMovement{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "dm-test",
+				Namespace: corev1.NamespaceDefault,
+			},
+			Spec: nnfv1alpha1.NnfDataMovementSpec{
+				Source: &nnfv1alpha1.NnfDataMovementSpecSourceDestination{
+					Path: srcPath,
+				},
+				Destination: &nnfv1alpha1.NnfDataMovementSpecSourceDestination{
+					Path: destPath,
+				},
+				UserId:  uint32(os.Geteuid()),
+				GroupId: uint32(os.Getegid()),
+				Cancel:  false,
+			},
+		}
+	})
+
+	JustBeforeEach(func() {
+		Expect(k8sClient.Create(context.TODO(), dm)).To(Succeed())
+	})
+
+	JustAfterEach(func() {
+		Expect(k8sClient.Delete(context.TODO(), dm)).To(Succeed())
+		Eventually(func() error {
+			return k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(dm), dm)
+		}).ShouldNot(Succeed())
+		Expect(os.Remove(srcPath)).To(Succeed())
+		Expect(os.Remove(destPath)).To(Succeed())
+
+	})
+
+	Context("when a data movement operation succeeds", func() {
+		It("should have a state and status of 'Finished' and 'Success'", func() {
+			Eventually(func(g Gomega) {
+				g.Expect(k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(dm), dm)).To(Succeed())
+				g.Expect(dm.Status.State).To(Equal(nnfv1alpha1.DataMovementConditionTypeFinished))
+				g.Expect(dm.Status.Status).To(Equal(nnfv1alpha1.DataMovementConditionReasonSuccess))
+			}, dmTestTimeout).Should(Succeed())
```

Comment:
While this isn't _wrong_, I think the general assertion strategy is to return the type you want to check; then use the Should() block to verify. (I'm actually not sure what `Should(Succeed())` does when the func returns nothing)

So this would look something like
```go
Eventually(func(g Gomega) nnfv1alpha1.NnfDataMovementStatus {
    g.Expect(k8sClient.Get(...)).To(Succeed())
    return dm.Status
}, "3s").Should(MatchFields(IgnoreExtras, Fields{
    "State": Equal(nnfv1alpha1.DataMovementConditionTypeFinished),
    "Status": Equal(nnfv1alpha1.DataMovementConditionReasonSuccess),
}))
```

Reply:
That makes way more sense.

Reply:
I can't figure out a way to mix `MatchFields` with `Eventually`. The `Should()` doesn't accept it.